File: 2025-02-February-GFG-POTD/February 22 - Longest valid Parentheses/Solution.cpp
// C++ Implementation
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxLength(string s) {
        stack<int> stack;
        stack.push(-1); // Base index for valid substring
        int maxLen = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.empty()) {
                    stack.push(i);
                } else {
                    maxLen = max(maxLen, i - stack.top());
                }
            }
        }
        return maxLen;
    }
};

int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S;
        Solution ob;
        cout << ob.maxLength(S) << endl;
        cout << "~" << endl;
    }
    return 0;
}

File: 2025-02-February-GFG-POTD/February 22 - Longest valid Parentheses/Solution.py
# Python Implementation
def maxLength(s):
    stack = [-1]  # Base index for valid substring
    max_len = 0

    for i in range(len(s)):
        if s[i] == '(':
            stack.append(i)
        else:
            stack.pop()
            if not stack:
                stack.append(i)
            else:
                max_len = max(max_len, i - stack[-1])
    
    return max_len

if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        S = input()
        print(maxLength(S))
        print("~")
